### ISSUE TITLE
Custom training data

### ISSUE THREAD

**Issue by Anand195** [flairNLP/flair#3022]
Hello, guys

Previously I was using space to train the NER model on my own custom dataset!
After I heard about the FLAIR NLP, I would like to try this!

Do let me know how to create custom training station for FLAIR NLP as well how I can train a model using my own custom dataset.

Any help is appreciable!

**Comment by Anand195** [comment]
Thanks for quick reply.
I have followed the official docs from Flair as you suggested but every time I try to train the model I got the f1-score value as 0.0 only.
Here is the screenshot of model training for your reference, also I the screenshot consisting training data.

!!!model training logs
<img width="1680" alt="Screenshot 2022-12-12 at 10 21 42 PM" src="https://user-images.githubusercontent.com/25666718/207232597-9ea3f0a1-64ed-467e-97cf-800b6de75625.png">

!!! Dataset screenshot!
<img width="247" alt="Screenshot 2022-12-13 at 10 45 43 AM" src="https://user-images.githubusercontent.com/25666718/207232654-3baf71cb-6906-49b6-837b-c2016f0feeb8.png">

**Comment by Anand195** [comment]
While searching for solution my dataset set in the format of BIO and I found this snippet

`tagger = SequenceTagger(hidden_size=256,
                        embeddings=embeddings,
                        tag_dictionary=label_dict,
                        tag_type=label_type,
                        tag_format="BIO",
                        use_crf=True)`

After running this I got this error 

` You can create a Dictionary that handles unknown items with an <unk>-key by setting add_unk = True in the construction.`

Do let me know how I can add this "add_unk - True" and where!

You support is valuable!

**Comment by Anand195** [comment]
Hey, @Guust-Franssens 
Thanks for sharing,
I have gone trough your solution and tried, all the spans are as expected but still can't get f1 score for and most of the time I'm getting BAD_EPOCHS!
<img width="1158" alt="Screenshot 2022-12-22 at 6 21 38 PM" src="https://user-images.githubusercontent.com/25666718/209138337-a5f017c6-3970-45ba-907a-373f1b2b0485.png">
Here is the screenshot for reference

Any help is appreciable !

**Comment by Guust-Franssens** [comment]
What is the tag format you have? BIO or BIOES?

Did you get around the error with regards to add_unk = True?

If you have issues with BIOES, could you try downgrading your Flair version to 0.10 and see if your model manages to learn anything in that version?

**Comment by Anand195** [comment]
Yes, I'm using the BIO tag format.

I strike on add_unk = True error, also after some research I managed to get around this error! 

surely I'll try to train the FLAIR by downgrading to 0.10!

Thanks for the quick response!

**Comment by Anand195** [comment]
While trying on using GPU instance for just 10 documents it shows CUDA_OUT_OF_MEMEORY!

<img width="1224" alt="Screenshot 2022-12-31 at 9 38 13 AM" src="https://user-images.githubusercontent.com/25666718/210124407-2e4edebb-c6b7-49ac-b911-aae15a18555c.png">


It took too much GPU memory for such a small dataset!
Can you help me to fix this?

**Comment by alanakbik** [comment]
Hello @Anand195 from your first screenshots it looks like your data set is quite small with only few update steps per epoch. This probably explains why the model is not learning anything. Can you print the corpus statistics (`print(corpus)`) and let us know the size of the dataset? 

Regarding the GPU memory issue it always depends on which transformer model you use. xlm-roberta-large for instance is a huge model and requires lots of memory. bert-base-uncased or distilbert-base-uncased are smaller and work also on smaller graphics cards.

**Comment by Anand195** [comment]
Hi @alanakbik ,
I have change the model from xlm-roberta-large to distilbert-base-uncased 

Here is the corpus output 

`=============== Corpus: 88 train + 88 dev + 10 test sentences ==================`

after performing model training for while got the CUDA error!
take a look.

<img width="1331" alt="Screenshot 2023-01-11 at 2 39 35 PM" src="https://user-images.githubusercontent.com/25666718/211764717-de418b35-6a31-40c8-a2e5-99c3b03a5a78.png">


if you nee here is the code snippet I use to train the model.

`
from flair.embeddings import WordEmbeddings, FlairEmbeddings, StackedEmbeddings
from flair.datasets import ColumnCorpus
import os
from flair.datasets import CONLL_03
from flair.embeddings import TransformerWordEmbeddings
from flair.models import SequenceTagger
from flair.trainers import ModelTrainer

# 1. get the corpus
# 1. get the corpus
columns = {0: 'text', 1: 'ner'}
# dataPath = "/Users/caypro/Downloads/Section-Data_1/flair/"
dataPath = "/content/"
corpus = ColumnCorpus(dataPath,columns,train_file="conllpp_train.txt",test_file="conllpp_test.txt",dev_file="conllpp_train.txt")
print(f"\n=============== {corpus} ================== \n")

# 3. make the label dictionary from the corpus
label_dict = corpus.make_label_dictionary(label_type='ner')
print(label_dict)

# 4. initialize fine-tuneable transformer embeddings WITH document context
embeddings = TransformerWordEmbeddings(model='distilbert-base-uncased',
                                       layers="-1",
                                       subtoken_pooling="first",
                                       fine_tune=True,
                                       use_context=True,
                                       )

# 5. initialize bare-bones sequence tagger (no CRF, no RNN, no reprojection)
tagger = SequenceTagger(hidden_size=256,
                        embeddings=embeddings,
                        tag_dictionary=label_dict,
                        tag_type='ner',
                        use_crf=False,
                        use_rnn=False,
                        reproject_embeddings=False,
                        )

# 6. initialize trainer
trainer = ModelTrainer(tagger, corpus)

# 7. run fine-tuning
trainer.fine_tune('/content/model',
                  max_epochs = 10,
                  learning_rate=5.0e-6,
                  mini_batch_size=4,
                  mini_batch_chunk_size=1,  # remove this parameter to speed up computation if you have a big GPU
                  )

`

Do let me know if any better solution is there!

**Comment by helpmefindaname** [comment]
Hi @Anand195 how long are your sentences? It could be possible, that they are really long and therefore a single sentence takes a lot of memory. If that is the case, splitting up the sentence into multiple should make it easier to process

**Comment by Anand195** [comment]
HI @helpmefindaname , thanks for the quick reply.
The sentences are not that long, the sentences are under 100-150 tokens length.
If you have any better solution to utilise flair to perform NER on custom dataset or share any example of that.

**Comment by trompx** [comment]
Hey @Anand195, did you solved it? Do you mind sharing what went wrong and if you managed to reproduce a 90+ F1 score on your custom small dataset?

**Comment by Kartikey-Aurva** [comment]
Hey @Anand195, Were you able to resolve this issue?
I am facing the exact same issue, where after every epoch I get a message "BAD EPOCH" and all F score, Recall, Precision everything is 0. 
I am training on MacOS and label format is BIO.
I tried with 2 type of dataset, one had 25000 sentences while other had just 100. I am facing same issue with both of them.
I tried changing mini batch size, learning rate but no use.

Results:
- F-score (micro) 0.0
- F-score (macro) 0.0
- Accuracy 0.0

By class:
              precision    recall  f1-score   support

PHONE_NUMBER     0.0000    0.0000    0.0000       6.0
         PER     0.0000    0.0000    0.0000       2.0

   micro avg     0.0000    0.0000    0.0000       8.0
   macro avg     0.0000    0.0000    0.0000       8.0
weighted avg     0.0000    0.0000    0.0000       8.0
